subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/AbstractRecompilationSpecProvider.java

```diff
@@ -117,6 +119,29 @@ private String rebuildClauseForChangedNonSourceFile(FileChange fileChange) {
         return String.format("%s '%s' has been %s", "resource", fileChange.getFile().getName(), fileChange.getChangeType().name().toLowerCase(Locale.US));
     }
 
+    /**
+     * For Groovy/Java joint compilation we treat recompilation of affected classes from Java source as an ABI change.
+     * This means we recompile all class dependencies, even if there was just a private change that affects class.
+     * This is because Groovy compiler tries to load these classes in some cases.
+     *
+     * Fix for issue <a href="https://github.com/gradle/gradle/issues/22531">#22531</a>.
+     */
+    private void processGroovyJavaJointCompilationAdditionalDependencies(JavaCompileSpec spec, RecompilationSpec recompilationSpec, SourceFileChangeProcessor sourceFileChangeProcessor, SourceFileClassNameConverter sourceFileClassNameConverter) {
+        if (!supportsGroovyJavaJointCompilation(spec)) {
+            return;
+        }
+        Set<String> classesWithJavaSource = recompilationSpec.getClassesToCompile().stream()
+            .flatMap(classToCompile -> sourceFileClassNameConverter.getRelativeSourcePaths(classToCompile).stream())
+            .filter(sourcePath -> sourcePath.endsWith(".java"))
+            .flatMap(sourcePath -> sourceFileClassNameConverter.getClassNames(sourcePath).stream())
+            .collect(Collectors.toSet());
+        if (!classesWithJavaSource.isEmpty()) {
+            sourceFileChangeProcessor.processChange(classesWithJavaSource, recompilationSpec);
```

Comment:
It feels like this method isn't idempotent. So if we feed the new classes to compile back into this method then we may even end up with a bigger change set, right? And I suppose we need the full closure here, so we'd need to do this until we don't add any more classes anymore I suppose.

You should be able to expose this in a test when you have a longer chain of private dependencies between Java classes, right?

My expectation was that we'd add all dependencies here when the source file is a Java file and we are in Groovy/Java joint compilation: https://github.com/gradle/gradle/blob/aa3807e513490867e6f2859ad2423a415513b74f/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java#L131

Reply:
I wanted to avoid adding logic there, since it's already complicated. We might have to add full closure, you are right. I will try to test what happens for classes that are loaded when current problematic class is loaded.

Reply:
After further investigation yes, there is still issue is case:
when we have:
```
api/A
impl/B1 -> class B1 { void m1() { A a = new A(); a.m1(); }; }
impl/C1 -> class C1 extends B1 { static boolean getCache() { return true; } }
impl/D1 -> class D1 { boolean isCacheEnabled = C1.cache }

impl/B2 -> class B2 { void m1() { B1 b = new B1(); b.m1(); }; }
impl/C2 -> class C2 extends B2 { static boolean getCache() { return true; } }
impl/D2 -> class D2 { boolean isCacheEnabled = C2.cache }
```

And we do a change:
```
api/A -> class A { void m1() {}; void m2() {}; }
impl/D1 -> class D1 { boolean isCacheEnabled = C1.cache; int a = 0; }
impl/D2 -> class D2{ boolean isCacheEnabled = C2.cache; int a = 0; }
```

Then we would get `NoClassDefException` for `B2`. 

After some thinking about it, I think the only change needed is, that we don't collect private dependents in that loop you commented for `.java` sources, but we collect only `public` dependents. We basically need to compile just all `subclasses` that are public dependents of some class. This is because if we look at:
```
impl/B2 -> class B2 { void m1() { B1 b = new B1(); b.m1(); }; }
impl/C2 -> class C2 extends B2 { static boolean getCache() { return true; } }
```
We have next cases:
- B1 is a private dependency of B2 and B1 doesn't have an ABI change: We don't need to recompile B2. And therefore we never delete B2, so loading class C2 always works
- B1 is a private dependency of B2 and B1 has ABI change: We need to recompile B2, but in that case B2 will be part of `classesWithJavaSource` so we will also recompile public dependents of B2: C2. So loading of C2 won't fail.
- B1 is a private dependency of B2 and B1 doesn't have an ABI change, but there is some other source change to B2 (either a source file change or a dependency change): We need to recompile B2, but in that case B2 will be part of `classesWithJavaSource` so we will also recompile public dependents of B2: C2. So loading of C2 won't fail.
- B1 is public dependency of B2: For any change to B1 in that case we will recompile B2. Since B1 is part of  `classesWithJavaSource` and so we will recompile public dependents of B1, and these are B2 and C2. So loading of C2 won't fail.

Reply:
I think that is not enough, back to drawing board.

Reply:
Seems like Groovy compilation still works in that case. So let's roll with the current set of changes.